addon/background.js

```diff
@@ -0,0 +1,207 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/* global browser */
+
+function log(...args) {
+  // console.log(...["[TESTPILOT v2] (background)"].concat(args)); // eslint-disable-lint no-console
+}
+
+const storage = browser.storage.local;
+const RESOURCE_UPDATE_INTERVAL = 4 * 60 * 60 * 1000; // 4 hours
+const TWO_WEEKS = 2 * 7 * 24 * 60 * 60 * 1000;
+
+/* browser action constants */
+const BROWSER_ACTION_LINK_BASE = [
+  "/experiments",
+  "?utm_source=testpilot-addon",
+  "&utm_medium=firefox-browser",
+  "&utm_campaign=testpilot-doorhanger"
+].join("");
+const BROWSER_ACTION_LINK_NOT_BADGED = BROWSER_ACTION_LINK_BASE +
+      "&utm_content=not+badged";
+const BROWSER_ACTION_LINK_BADGED = BROWSER_ACTION_LINK_BASE +
+      "&utm_content=badged";
+let BROWSER_ACTION_LINK = BROWSER_ACTION_LINK_NOT_BADGED;
+
+const resources = {
+  experiments: [],
+  news_updates: []
+};
+
+let currentEnvironment = {
+  name: "production",
+  baseUrl: "https://testpilot.firefox.com"
+};
+
+function uuidv4() {
+  return ([1e7]+-1e3+-4e3+-8e3+-1e11).replace(/[018]/g, c => // eslint-disable-line space-infix-ops
+                                              (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16));
+}
+
+function setInstalledTxpAddons() {
+  return browser.management.getAll().then((infoArray) => {
+    const installed = infoArray.map((exp) => exp.id);
+    const txpAddons = resources.experiments.map((exp) => exp.id);
+
+    return installed.filter(function(n) {
+      return txpAddons.indexOf(n) !== -1;
+    });
+  });
+}
+
+async function setup() {
+  const { clientUUID } = await storage.get("clientUUID");
+  if (!clientUUID) {
+    await storage.set({ clientUUID: uuidv4() });
+  }
+  setupEnvironment();
+  setupBrowserAction();
+  await fetchResources();
+  setDailyPing();
+}
+
+async function setDailyPing() {
+  const delayInMinutes = 1;
+  const periodInMinutes = 60; // check hourly
+  const { lastPing } = await storage.get("lastPing");
+  let initial = false;
+
+  if (!lastPing) {
+    initial = true;
+    await storage.set({ lastPing: Date.now()});
+  }

```

Comment:
we can remove `initial` and skip this if we default `lastPing` to yesterday

Reply:
woah clever :+1: